nix/pyproject.toml

```
@@ -12,18 +12,20 @@ python = "^3.8"
 appdirs = "^1.4.4"
 argcomplete = "^1.12.2"
 colorama = "^0.4.4"
-hjson = "^3.0.2"
-milc = "^1.1.0"
-Pygments = "^2.8.0"
 dotty-dict = "^1.3.0"
+hjson = "^3.0.2"
 jsonschema = "^3.2.0"
+milc = "^1.3.0"
+Pygments = "^2.8.0"
+qmk = "^0.0.46"
```

Current version is now 0.0.48, apparently.

---

That should probably be a wildcard dependency (unless there's a hard minimum version) since we'll always want the latest:

```suggestion
qmk = "*"
```

But note that when the version changes upstream, it won't have any automatic effect: `poetry update --lock` is needed in order to re-calculate and lock the set of package versions into `poetry.lock`, at which point the dependency specs in `pyproject.toml` are re-applied to the latest packages in PyPi. @sigprof: it's probably worth doing that now, with the change suggested above.

---

(More on the syntax for dependency specs here: https://python-poetry.org/docs/dependency-specification/)

---

Should I just run  `poetry update --lock` after changing the dependency for `qmk`, bumping the locked revisions for all other Python packages too?

And the compatibility constraints here can be complicated — e.g., I know that `milc` 1.3.0 (or maybe a version just before it) broke compatibility with older QMK code by removing a previously exported function.

---

Yep, if the contraints in `pyproject.toml` accurately represent what the code here is compatible with, then `poetry update --lock` is the thing to do: it will bump everything possible to the latest compatible versions, similar to what you'd get with `pip`-installing, but locking the results for reproducibility. So probably worth checking all the constraints in `pyproject.toml`, e.g. `milc`. Wildcard or minimum versions are probably best.

---

> (unless there's a hard minimum version)

The hard minimum version should probably be the latest release of the `qmk` Python package at the time of the latest `qmk_firmware` commit (so it would need regular updates). Or we could just put `*` there and just bump the locked versions whenever `qmk/qmk_cli` gets a new release (sometimes a too new `qmk_cli` can break compatibility with older `qmk_firmware` versions too, so locking is actually useful here).

The actual working of Python dependencies here is somewhat strange — if you use the `qmk` executable from `$PATH` (so that it comes from the `qmk` Python package, which is `qmk/qmk_cli` actually), that Python code will actually find the `qmk_firmware` directory and import Python modules from `qmk_firmware/lib/python`, using the Python module path which in the Nix case was set up by the Nix-generated wrapper for the `qmk` executable. This works, because `qmk_cli` has dependencies on all Python modules that could be used by the code in `qmk_firmware`, even though the code in `qmk_cli` does not actually use most of those modules; this also may be the reason to need the updated `qmk_cli` version (if some new code with new Python requirements was added to `qmk_firmware`, it won't work until you get a new `qmk_cli` version which has those modules available in its Python module path). The rest of Python environment generated by `poetry2nix.mkPoetryEnv` is only needed if you call `bin/qmk` directly, so that the `python3` executable from that environment is used.

---

> Or we could just put `*` there and just bump the locked versions whenever `qmk/qmk_cli` gets a new release

That's what I would suggest.

> The actual working of Python dependencies here is somewhat strange

I think that adding the dependency on `qmk` should cause the `qmk_cli` on the `$PATH` inside a nix shell to be the correct one, right? Since we're depending on `qmk`, all of its dependencies will get merged with any additional ones in the `pyproject.toml` here, so I think things should work out. Maybe you're suggesting that we could just depend on `qmk`? That seems likely.